tests/functional/utils.py

```diff
@@ -96,3 +96,19 @@ def fail_json(*args, **kwargs):
 
     kwargs["failed"] = True
     raise AnsibleFailJson(args, kwargs)
+
+
+def side_effects_with_exceptions(side_effects):
+    """
+    Works similarly to `MagicMock(side_effect=side_effects)` as if side_effects was a list,
+    but if item in the list is instance of an exception, it raises it instead of returning it.
+    """
+    side_effects = side_effects.copy()
+
+    def _exec_side_effect(*args, **kwargs):
+        i = side_effects.pop(0)
+        if isinstance(i, Exception):
+            raise i
```

Comment:
Why to raise the error? It should be sufficient to return it as side effect so the mock can used it. The error should be raised in the mocked function, not here in the wrapper.

Reply:
Also, I would rename it from `_exec_side_effect` to `_pop_side_effect` or something similar, as the side effect execution should be handled by pytest, here we just use nice wrapper to return the side effects.

Reply:
To raise the exception is essentially the reason for existence of this wrapper. In MagicMock side_effect you can choose from two different modes of operation, either you pass an exception and it raises it, or you can pass a list of values and it returns the values one by one, without raising. It does not combine those two effects, which is exactly what I need to test certain setups, e.g. exception raised on second update, while the first one passes

Reply:
Thanks for explanation. Then I have no more objections :)

Reply:
Renamed to your suggestion